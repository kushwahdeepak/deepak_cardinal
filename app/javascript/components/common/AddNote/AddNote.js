import React, { useState, useRef, useEffect } from 'react'
import Image from 'react-bootstrap/Image'
import axios from 'axios'

import styles from './styles/AddNote.module.scss'
import { makeRequest } from '../RequestAssist/RequestAssist'
import { MentionsInput, Mention } from 'react-mentions'
import PencilIcon from '../../../../assets/images/icons/pencil-icon-v2.svg'

function AddNote({ candidate, setNoteAdded }) {
    const [note, setNote] = useState('')
    const [expandNote, setExpandNote] = useState(false)
    const noteInputAreaRef = useRef()
    const [mentionedMembers, setMentionedMembers] = useState([])
    const [members, setMembers] = useState([])
    const [loading, setLoading] = useState(false)

    const addNote = () => {
        setLoading(true)
        var token = document.querySelector('meta[name="csrf-token"]').content
        axios
            .post(`/people/${candidate.id}/notes`, {
                authenticity_token: token,
                body: note,
                person_id: candidate.id,
                mention_ids: mentionedMembers.map((m) => m.id),
            })
            .then((res) => {
                setNoteAdded(true)
                setNote('')
                setExpandNote(false)
                setLoading(false)
            })
            .catch((err) => {
                console.log(err)
                setLoading(false)
            })
    }

    const getWrapperStyle = () => {
        return expandNote ? { height: 'max-content' } : { height: '30px' }
    }

    const expandInput = () => {
        setExpandNote(!expandNote)
    }

    useEffect(() => {
        if (noteInputAreaRef.current && expandNote)
            noteInputAreaRef.current.focus()
    }, [expandNote])

    const getOrganizationMembers = async () => {
        const url = '/get_users_in_organization'
        const result = await makeRequest(url, 'get', {})
        if (result?.data.member.length == 0) {
            setMembers([{ id: -1, display: 'No members to mention' }])
        } else {
            setMembers(
                result?.data.member.map((m) => ({
                    id: m.id,
                    display: `${m.first_name + ' ' + m.last_name}`,
                    email: '',
                }))
            )
        }
    }

    const onAdd = (id) => {
        if (id == -1) return
        setMentionedMembers([
            ...mentionedMembers,
            members.find((m) => m.id === id),
        ])
    }

    const removeMention = () => {
        setMentionedMembers(
            mentionedMembers.filter((m) => note.includes(`@[${m.display}]`))
        )
    }

    useEffect(() => {
        removeMention()
    }, [note])

    useEffect(() => {
        getOrganizationMembers()
    }, [])

    const defaultStyle = {
        control: {
            fontSize: 12,
        },

        highlighter: {
            overflow: 'hidden',
        },

        input: {
            margin: 0,
        },

        '&singleLine': {
            control: {
                display: 'inline-block',
                width: 130,
            },

            highlighter: {
                padding: 1,
                border: '2px inset transparent',
            },

            input: {
                padding: 10,
                border: '2px inset',
            },
        },

        '&multiLine': {
            control: {
                fontFamily: 'inherit',
            },

            highlighter: {
                padding: 9,
            },

            input: {
                padding: 9,
                minHeight: 20,
                outline: 0,
                border: 0,
            },
        },

        suggestions: {
            list: {
                backgroundColor: 'white',
                border: '1px solid rgba(0,0,0,0.15)',
                fontSize: 10,
            },

            item: {
                borderBottom: '1px solid rgba(0,0,0,0.15)',

                '&focused': {
                    backgroundColor: '#cee4e5',
                },
            },
        },
    }

    const defaultMentionStyle = {
        backgroundColor: '#cee4e5',
    }

    return (
        <>
            {expandNote && (
                <div
                    className={styles.backdrop}
                    onClick={() => setExpandNote(false)}
                ></div>
            )}
            <div className={styles.addNotesTextarea} style={getWrapperStyle()}>
                <div
                    className="position-relative"
                    style={{
                        height: '100%',
                        maxHeight: '30px',
                        background: '#4c68ff',
                    }}
                >
                    <Image
                        src={PencilIcon}
                        style={{
                            position: 'absolute',
                            height: '10px',
                            width: '10px',
                            bottom: '8px',
                            left: '10px',
                        }}
                    />
                    <input
                        placeholder="Add Notes"
                        type="text"
                        onClick={expandInput}
                    />
                </div>
                <div
                    style={{
                        display: expandNote ? 'flex' : 'none',
                        flexDirection: 'column',
                        paddingLeft: '7px',
                        paddingRight: '7px',
                    }}
                >
                    <input
                        type="text"
                        disabled
                        placeholder="@ mention to share with a coworker (optional)"
                        className={styles.mentionInput}
                    />

                    <MentionsInput
                        inputRef={noteInputAreaRef}
                        value={note}
                        onChange={(e) => setNote(e.target.value)}
                        style={defaultStyle}
                    >
                        <Mention
                            trigger="@"
                            data={members}
                            markup="@[__display__]"
                            displayTransform={(id, display) => '@' + display}
                            renderSuggestion={(
                                suggestion,
                                search,
                                highlightedDisplay,
                                index,
                                focused
                            ) => {
                                const [firstName, lastName] = suggestion.display
                                    ? suggestion.display.split(' ')
                                    : ['', '']
                                return (
                                    <div
                                        className={styles.memberRow}
                                        onClick={
                                            suggestion.id == -1
                                                ? (e) => {
                                                      e.stopPropagation()
                                                      e.preventDefault()
                                                  }
                                                : () => {}
                                        }
                                    >
                                        {suggestion.id == -1 ? (
                                            <p>No members to display</p>
                                        ) : (
                                            <>
                                                <div className={styles.logo}>
                                                    {firstName
                                                        ? firstName
                                                              .charAt(0)
                                                              .toUpperCase()
                                                        : ''}
                                                    {lastName
                                                        ? lastName
                                                              .charAt(0)
                                                              .toUpperCase()
                                                        : ''}
                                                </div>
                                                <div
                                                    style={{
                                                        marginLeft: '15px',
                                                    }}
                                                >
                                                    <p className={styles.text}>
                                                        @{firstName}
                                                    </p>
                                                    <p className={styles.text}>
                                                        {highlightedDisplay}
                                                    </p>
                                                </div>
                                            </>
                                        )}
                                    </div>
                                )
                            }}
                            onAdd={onAdd}
                            style={defaultMentionStyle}
                        />
                    </MentionsInput>

                    <button
                        className={styles.doneButton}
                        onClick={addNote}
                        disabled={loading}
                    >
                        {loading ? 'Submitting...' : 'Done'}
                    </button>
                </div>
            </div>
        </>
    )
}

export default AddNote
