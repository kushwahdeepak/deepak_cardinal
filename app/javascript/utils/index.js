import { act } from 'react-dom/test-utils'

export const findByTestAttr = (wrapper, attr) => {
    return wrapper.find(`[data-test='${attr}']`)
}

export const timeoutResolve = (time = 0, response = {}) =>
    new Promise((resolve) => {
        setTimeout(() => resolve(response), time)
    })

export const wait = (time = 0) => {
    const timer = timeoutResolve(time)

    return act(async () => {
        await timer
    })
}
export function emailIsValid(email) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)
}

export function generateRandom(min, max) {
    return Math.floor(min + Math.random() * (max + 1 - min))
}

export function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase()
}

export function createMarkedList(array) {
    if (!Array.isArray(array)) {
        return []
    }
    return array.map((candidate) => ({
        id: candidate.id,
        name: candidate.id,
        fullName: `${candidate.first_name} ${candidate.last_name}`,
        emailAddress: candidate.email_address,
        isChecked: false,
    }))
}

export function stringReplace(string,toSymbol,fromSymbol){
    return string.replace(toSymbol, fromSymbol)
}

export function capitalize(string) {
    return string.replace(/\b(\w)/g, s => s.toUpperCase());
}

export function skillIsNotEmpty(value){
    return value.trim().length > 0;
}

export function firstCharacterCapital(value){
    return value?.charAt(0)?.toUpperCase() + value?.slice(1)
}
