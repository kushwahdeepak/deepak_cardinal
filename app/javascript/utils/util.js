import isNil from 'lodash.isnil'
import isEmpty from 'lodash.isempty'

class Util {
    /**
     * Define stage values
     */
    static STAGES = ['first_interview', 'second_interview', 'offer', 'reject']

    /**
     * Tests whether a single object attribute includes the at
     * least one string from the given set.
     */
    static attributePassesStringSetFilter(object, attr, stringSet) {
        if (isNil(object[attr])) return false

        // We "OR" the tag tests: the attribute needs to contain only one tag
        return stringSet
            .map((string) => string.trim().toLowerCase())
            .reduce(
                (passes, string) =>
                    passes ||
                    Util.attributeToString(object[attr])
                        .toLowerCase()
                        .includes(string),
                false
            )
    }

    /**
     * Tests whether at least one of a set of object attributes
     * includes every word from a given piece of text.
     */
    static objectPassesTextFilter(object, attrs, text) {
        text = text.trim().toLowerCase()

        if (text === '') return true

        const words = text.split(' ')

        // Join all the given attribute names to lowercase, stringified attribute values
        const allAttrsString = attrs
            .filter((attr) => !isNil(object[attr]))
            .map((attr) => Util.attributeToString(object[attr]).toLowerCase())
            .join('')

        // Is every word included in the allAttrsString?
        return words.reduce(
            (matchesAll, word) => matchesAll && allAttrsString.includes(word),
            true
        )
    }

    static attributeToString(attr) {
        if (typeof attr === 'string') {
            return attr
        } else if (Array.isArray(attr)) {
            return attr.join(',')
        } else if (typeof attr === 'object') {
            return JSON.stringify(attr)
        } else {
            return '' + attr
        }
    }

    static numberWithCommas(x) {
        return x?.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
    }

    /**
     * Returns number of results in format [x] - [y] of [z]
     * @param {number} totalJobsCount Total number of jobs
     * @param {number} pageCount Total number of pages
     * @param {number} activePage Currently active page
     * @param {number} pageSize How many results to display per page
     */
    static displayNumberOfResults(
        totalJobsCount,
        pageCount,
        activePage,
        pageSize,
        totalPersonCount=null,
        candidatePage=false,
        showdisplaycount=true
    ) {
        if (totalJobsCount == 0) return '0 - 0 of 0'
        const offset = pageSize * pageCount - totalJobsCount
        const lastJobIndex =
            (activePage + 1) * pageSize > totalJobsCount
                ? totalJobsCount
                : (activePage + 1) * pageSize
        const firstJobIndex =
            activePage == 0
                ? 1
                : activePage == pageCount - 1
                ? lastJobIndex - pageSize + offset + 1
                : lastJobIndex - pageSize + 1
        return `${firstJobIndex} - ${lastJobIndex} of ${Util.numberWithCommas(
            (candidatePage && showdisplaycount) ? totalPersonCount : totalJobsCount

        )}`
    }

    /**
     * Returns a string indicating the candidate's title and current
     * company, based on available information on the Person object
     * @param {object} candidate Person object from our db
     */
    static candidateCompanyString(candidate) {
        const position = candidate.company_position
        const currentCompany = candidate.current_company

        if (isEmpty(position) && isEmpty(currentCompany)) return ''

        if (!isEmpty(position) && !isEmpty(currentCompany))
            return `${position} at ${currentCompany}`

        if (!isEmpty(position)) return position

        if (!isEmpty(currentCompany)) return `at ${currentCompany}`
    }

    /**
     * Converts the given array of objects with ids to an object whose
     * keys are the ids of the array objects.
     *
     * Every object in the given array must have a unique 'id' property defined.
     *
     * @param {Array} array
     * @returns {Object}
     */
    static arrayToObject(array) {
        const obj = {}
        array.forEach((element) => (obj[element.id] = element))
        return obj
    }

    /**
     * Makes a copy of the given array and replace the value at 'index' with 'value'
     *
     * @param {Array} array
     * @param {Number} index
     * @param {*} value
     * @returns {Array}
     */
    static replaceInArrayCopy_i(array, index, value) {
        const clone = array.slice()
        clone[index] = value
        return clone
    }

    /**
     * Makes a copy of the given array and replaces the element with the same
     * .id property as obj.id with obj
     *
     * @param {Array} array
     * @param {Object} obj
     * @returns {Array}
     */
    static replaceInArrayCopy(array, obj) {
        const index = array.findIndex((element) => element.id === obj.id)
        return Util.replaceInArrayCopy_i(array, index, obj)
    }

    static deleteInArrayCopy(array, index) {
        const copy = array.slice()
        copy.splice(index, 1)

        return copy
    }

    static handleUndefined(value){
        if (value === undefined || value === 'undefined')
        return '';
        else
        return value
    }

    static handleUndefinedFullName(firstName, lastName){
      if(firstName === "undefined" || firstName === undefined)
      return lastName
      else if(lastName === "undefined" || lastName === undefined)
      return firstName
      else
      return firstName + ' ' + lastName     }

    static capitalize(value){
         return value.charAt(0).toUpperCase() + value.slice(1).toLowerCase();
    }
}

export default Util
