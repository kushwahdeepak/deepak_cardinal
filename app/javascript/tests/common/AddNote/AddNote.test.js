import React from 'react';
import AddNote from '../../../components/common/AddNote/AddNote';
import renderer from 'react-test-renderer';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { findByTestAttr, wait } from '../../../utils';

configure({adapter: new Adapter()});

const initialShallowRender = (props = {}) => {
	return shallow(<AddNote {...props}/>);
};

const initialMountRender = (props = {}) => {
	return mount(<AddNote {...props}/>);
};

const initialProps = {
};

describe('AddNote component',  () => {
	let wrapper;
	let fullWrapper;

	beforeEach(() => {

		const props = {
			...initialProps,
			show: true
		};

		wrapper = initialShallowRender(props);
		fullWrapper = initialMountRender(props);
	});

	afterEach(() => {
		fullWrapper && fullWrapper.unmount();
	});

	it('should render AddNote component', () => {
		expect(wrapper.exists()).toBe(true);
	});

	it('should has modal wrapper', () => {
		expect(findByTestAttr(wrapper, 'modal-wrapper').length).toBe(1);
	});

	it('should match with a snapshot', () => {
		const component = renderer.create(<AddNote {...initialProps} />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe('AddNote component with props', () => {
		let wrapper;
		let fullWrapper;

		beforeEach(() => {

			const props = {
				...initialProps,
				onHide: jest.fn(),
				show: true
			};

			wrapper = initialShallowRender(props);
			fullWrapper = initialMountRender(props);
		});

		afterEach(() => {
			jest.clearAllMocks();
		});

		it('should call close click function once', () => {
			const closeButton = findByTestAttr(fullWrapper, 'close-button').at(0);
			closeButton.simulate('click');
			expect(fullWrapper.props().onHide).toHaveBeenCalledTimes(1);
		});

		it('should have the expected content', async () => {
			const addButton = findByTestAttr(fullWrapper, 'add-modal-btn').at(0);
			const closeButton = findByTestAttr(fullWrapper, 'close-button').at(0);
			const modalWrapper = findByTestAttr(fullWrapper, 'modal-wrapper').at(0);
			expect(addButton.exists()).toBeTruthy();
			expect(closeButton.exists()).toBeTruthy();
			expect(modalWrapper.exists()).toBeTruthy();
		});

		it('should click on the button', async () => {
			const closeButton = () => findByTestAttr(fullWrapper, 'close-button').at(0);
			closeButton().simulate('click');
			await wait(500);
			fullWrapper.update();

			const modalWrapper = () => findByTestAttr(fullWrapper, 'modal-wrapper').at(0);
			expect(modalWrapper()).toEqual({});
		});

		it.skip('should handle the correct value in textarea', () => {
			const textAreaItem = findByTestAttr(fullWrapper, 'text-area-item').at(0).find('textarea');
			textAreaItem.simulate('change', {
				value: 'Mock value'
			});
			const updatedTextArea = findByTestAttr(fullWrapper, 'text-area-item').at(0).find('textarea');
			expect(updatedTextArea.props().value).toBe('Mock value');
		});
	});
	describe('AddNote component without props', () => {
		let wrapper;
		let fullWrapper;

		beforeEach(() => {

			const props = {
				...initialProps,
				onHide: jest.fn(),
				show: false
			};

			wrapper = initialShallowRender(props);
			fullWrapper = initialMountRender(props);
		});
	});

});
