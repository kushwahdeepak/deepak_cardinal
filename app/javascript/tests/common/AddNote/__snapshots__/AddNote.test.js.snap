// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNote component should match with a snapshot 1`] = `null`;
